### PR TITLE
Refactoring the FCM API

### PR DESCRIPTION
**Pull request by hiranya911**
This PR implements some design changes in order to improve the modularity and testability of the `FirebaseMessaging` implementation.

* Turned `FirebaseMessagingClient` and `InstanceIdClient` into interfaces.
* Implementations of these interfaces are now injected into `FirebaseMessaging` via suppliers. This makes it possible to unit test the `FirebaseMessaging` API surface by mocking out the dependencies.
* Re-org'ed the existing tests, and added a number of new test cases to bring the test coverage of core classes to 100%.

### CODE REVIEW

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+    MessagingBatchCallback callback = new MessagingBatchCallback();
+    BatchRequest batch = newBatchRequest(messages, dryRun, callback);
+    batch.execute();
+    return new BatchResponse(callback.getResponses());
+  }
+
+  private BatchRequest newBatchRequest(
+      List<Message> messages, boolean dryRun, MessagingBatchCallback callback) throws IOException {
+
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java]: Removed the extra whitespace. I'll keep the comment for now. It is really reminder for me :)

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+  private FirebaseMessagingClientImpl.Builder fullyPopulatedBuilder() {
+    return FirebaseMessagingClientImpl.builder()
+        .setProjectId("test-project")
+        .setJsonFactory(Utils.getDefaultJsonFactory())
+        .setRequestFactory(Utils.getDefaultTransport().createRequestFactory())
+        .setChildRequestFactory(Utils.getDefaultTransport().createRequestFactory());
+  }
+
+  private static Map<Message, Map<String, Object>> buildTestMessages() {
+    ImmutableMap.Builder<Message, Map<String, Object>> builder = ImmutableMap.builder();
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Done

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        HttpRequestInitializer initializer = requestFactory.getInitializer();
+        if (initializer != null) {
+          initializer.initialize(request);
+        }
+        request.setResponseInterceptor(responseInterceptor);
+      }
+    };
+  }
+
+  static final class Builder {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java]: I do see your point. However, I do feel that 4-5 are too many arguments for a single constructor. Will keep this tradeoff in mind for the future.

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 4]
```diff
@@ -0,0 +1,828 @@
...
+  @Test
+  public void testSend() throws Exception {
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    FirebaseMessagingClient messaging = initMessagingClient(response, interceptor);
+    Map<Message, Map<String, Object>> testMessages = buildTestMessages();
+
+    for (Map.Entry<Message, Map<String, Object>> entry : testMessages.entrySet()) {
+      response.setContent(MOCK_RESPONSE);
+      String resp = messaging.send(entry.getKey(), false);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Added some constants with human readable names (DRY_RUN_ENABLED, DRY_RUN_DISABLED)

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 5]
```diff
@@ -0,0 +1,828 @@
...
+  @Test
+  public void testSendAll() throws Exception {
+    final TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    FirebaseMessagingClient messaging = initMessagingClientForBatchRequests(
+        MOCK_BATCH_SUCCESS_RESPONSE, interceptor);
+    List<Message> messages = ImmutableList.of(
+        EMPTY_MESSAGE, EMPTY_MESSAGE
+    );
+
+    BatchResponse responses = messaging.sendAll(messages, false);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Added constants.

File: src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+          public InstanceIdClient get() {
+            return new InstanceIdClient(app, null);
+          }
+        })
+        .build();
+  }
+
+  private static class FailingHttpTransport extends HttpTransport {
+    @Override
+    protected LowLevelHttpRequest buildRequest(String method, String url) throws IOException {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java]: Done

File: src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(new MockGoogleCredentials("test-token"))
+        .setProjectId("test-project")
+        .build();
+    FirebaseApp app = FirebaseApp.initializeApp(options);
+
+    try {
+      InstanceIdClientImpl client = InstanceIdClientImpl.fromApp(app);
+
+      assertSame(options.getJsonFactory(), client.getJsonFactory());
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java]: I wanted to add some spacing between the method being tested (fromApp), and the assertions.

File: src/test/java/com/google/firebase/testing/TestUtils.java [hunk 8]
```diff
@@ -142,4 +143,13 @@ public static HttpRequest createRequest(MockLowLevelHttpRequest request) throws
     HttpRequestFactory requestFactory = transport.createRequestFactory();
     return requestFactory.buildPostRequest(TEST_URL, new EmptyContent());
   }
+
+  public static HttpTransport faultyHttpTransport() {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/testing/TestUtils.java]: Done

File: src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        assertEquals("test error", error.getMessage());
+        assertTrue(error.getCause() instanceof HttpResponseException);
+      }
+
+      checkTopicManagementRequestHeader(interceptor.getLastRequest(), TEST_IID_UNSUBSCRIBE_URL);
+    }
+  }
+
+  @Test
+  public void testUnsubscribeUnknownError() throws Exception {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java]: Done

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+      response.setStatusCode(code).setContent(
+          "{\"error\": {\"status\": \"INVALID_ARGUMENT\", \"message\": \"test error\", "
+              + "\"details\":[{\"@type\": \"type.googleapis.com/google.firebase.fcm"
+              + ".v1.FcmError\", \"errorCode\": \"UNREGISTERED\"}]}}");
+
+      try {
+        messaging.send(EMPTY_MESSAGE, false);
+        fail("No error thrown for HTTP error");
+      } catch (FirebaseMessagingException error) {
+        assertEquals("registration-token-not-registered", error.getErrorCode());
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Done

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+      SendResponse sendResponse = responses.get(i);
+      assertTrue(sendResponse.isSuccessful());
+      assertEquals("projects/test-project/messages/" + (i + 1), sendResponse.getMessageId());
+      assertNull(sendResponse.getException());
+    }
+    checkBatchRequestHeader(interceptor.getLastRequest());
+    checkBatchRequest(interceptor.getLastRequest(), 2);
+  }
+
+  private void assertSendBatchFailure(
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Done

File: src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    InstanceIdClient messaging = initMessaging(response, interceptor);
+    for (int statusCode : HTTP_ERRORS) {
+      response.setStatusCode(statusCode).setContent("{\"error\": \"test error\"}");
+      try {
+        messaging.subscribeToTopic("test-topic", ImmutableList.of("id1", "id2"));
+        fail("No error thrown for HTTP error");
+      } catch (FirebaseMessagingException error) {
+        assertEquals(getTopicManagementErrorCode(statusCode), error.getErrorCode());
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java]: Done

File: src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    FirebaseMessaging messaging = initMessaging(response, interceptor);
+    for (int statusCode : HTTP_ERRORS) {
+      response.setStatusCode(statusCode).setContent("{\"error\": \"test error\"}");
+      try {
+        messaging.subscribeToTopicAsync(ImmutableList.of("id1", "id2"), "test-topic").get();
+        fail("No error thrown for HTTP error");
+      } catch (ExecutionException e) {
+        assertTrue(e.getCause() instanceof FirebaseMessagingException);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientTest.java]: Done

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java [hunk 14]
```diff
@@ -0,0 +1,828 @@
...
+    );
+
+    BatchResponse responses = messaging.sendAll(messages, false);
+
+    assertSendBatchFailure(responses, interceptor);
+  }
+
+  @Test
+  public void testSendAllHttpError() {
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java]: Done

**Comment by hiranya911** [comment]
Thanks @schmidt-sebastian. Made the suggested changes, and ready for another round of feedback.

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+      setCommonFcmHeaders(request.getHeaders());
+      batch.queue(
+          request, MessagingServiceResponse.class, MessagingServiceErrorResponse.class, callback);
+    }
+    return batch;
+  }
+
+  private void setCommonFcmHeaders(HttpHeaders headers) {
+    headers.set("X-GOOG-API-FORMAT-VERSION", "2");
+    headers.set("X-Firebase-Client", clientVersion);
```
Comment (schmidt-sebastian) [inline src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java]: Nit: Declare these as constants.

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java [hunk 16]
```diff
@@ -1,82 +1,40 @@
-/*
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java]: Can you re-add this license comment? Thanks!

File: src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java [hunk 17]
```diff
@@ -0,0 +1,384 @@
+package com.google.firebase.messaging;
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java]: License header missing.

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClient.java [hunk 18]
```diff
@@ -1,243 +1,32 @@
...
-        MessagingServiceErrorResponse error, HttpHeaders responseHeaders) {
-      responses.add(SendResponse.fromException(newException(error)));
-    }
+interface FirebaseMessagingClient {
+
+  /**
+   * Sends the given message with FCM.
+   *
+   * @param message A non-null {@link Message} to be sent.
+   * @param dryRun a boolean indicating whether to perform a dry run (validation only) of the send.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/FirebaseMessagingClient.java]: Done

File: src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java (hunk 16 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java]: Done

File: src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java (hunk 17 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/messaging/InstanceIdClientImplTest.java]: Done

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java (hunk 15 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/FirebaseMessagingClientImpl.java]: Done

**Pull request merged**